solver/src/settlement_submission/retry.rs

```diff
@@ -66,18 +69,33 @@ impl<'a> TransactionSending for SettlementSender<'a> {
     type Output = SettleResult;
 
     async fn send(&self, gas_price: f64) -> Self::Output {
+        debug_assert!(!self.nodes.is_empty());
         tracing::info!(
             "submitting solution transaction at gas price {:.2} GWei",
             gas_price / 1e9
         );
-        let mut method =
-            settle_method_builder(self.contract, self.settlement.clone(), self.account.clone())
-                .nonce(self.nonce)
-                .gas_price(GasPrice::Value(U256::from_f64_lossy(gas_price)))
-                .gas(U256::from_f64_lossy(self.gas_limit));
-        method.tx.resolve = Some(ResolveCondition::Confirmed(ConfirmParams::mined()));
-        let result = method.send().await.map(|tx| tx.hash());
-        SettleResult(result)
+        let send_with_node = |node| {
+            let contract = GPv2Settlement::at(node, self.contract.address());
+            let mut method =
+                settle_method_builder(&contract, self.settlement.clone(), self.account.clone())
+                    .nonce(self.nonce)
+                    .gas_price(GasPrice::Value(U256::from_f64_lossy(gas_price)))
+                    .gas(U256::from_f64_lossy(self.gas_limit));
+            method.tx.resolve = Some(ResolveCondition::Confirmed(ConfirmParams::mined()));
+            async { SettleResult(method.send().await.map(|tx| tx.hash())) }.boxed()
+        };
+        let mut futures = self.nodes.iter().map(send_with_node).collect::<Vec<_>>();
+        loop {
+            let (result, _index, rest) = futures::future::select_all(futures).await;
+            match &result.0 {
+                Ok(_) => return result,
+                Err(_) if rest.is_empty() => return result,
+                Err(err) => {
+                    tracing::warn!(?err, "single node tx failed");
```

Comment:
Now that we have multiple web3 instances, maybe it also makes sense for the `Transport` implementation to log some additional information like an index or the node URL to make debugging the transport logs a little easier.

Reply:
Logging the index here is complicated because we lose the initial order of the futures after going through select_all. Will make another PR for Transport to log this too.